assets/src/js/bindings/block-editor.js

```diff
@@ -0,0 +1,407 @@
+	InspectorControls,
+	useBlockBindingsUtils,
+} from '@wordpress/block-editor';
+import {
+	ComboboxControl,
+	__experimentalToolsPanel as ToolsPanel,
+	__experimentalToolsPanelItem as ToolsPanelItem,
+} from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+import { useSelect } from '@wordpress/data';
+import { store as coreDataStore } from '@wordpress/core-data';
+import { store as editorStore } from '@wordpress/editor';
+
+// These constant and the function above have been copied from Gutenberg. It should be public, eventually.
+
+const BLOCK_BINDINGS_ALLOWED_BLOCKS = {
+	'core/paragraph': [ 'content' ],
+	'core/heading': [ 'content' ],
+	'core/image': [ 'id', 'url', 'title', 'alt' ],
+	'core/button': [ 'url', 'text' ],
+};
+
+const BLOCK_BINDINGS_RELATED_FIELD_TYPES = {
+	'core/paragraph': {
+		content: [ 'text', 'textarea', 'date_picker', 'number', 'range' ],
+	},
+	'core/heading': {
+		content: [ 'text', 'textarea', 'date_picker', 'number', 'range' ],
+	},
+	'core/image': {
+		id: [ 'image' ],
+		url: [ 'image' ],
+		title: [ 'image' ],
+		alt: [ 'image' ],
+	},
+	'core/button': {
+		url: [ 'url' ],
+		text: [ 'text', 'checkbox', 'select', 'date_picker' ],
+		linkTarget: [ 'text', 'checkbox', 'select' ],
+		rel: [ 'text', 'checkbox', 'select' ],
+	},
+};
+
+/**
+ * Gets the bindable attributes for a given block.
+ *
+ * @param {string} blockName The name of the block.
+ *
+ * @return {string[]} The bindable attributes for the block.
+ */
+function getBindableAttributes( blockName ) {
+	return BLOCK_BINDINGS_ALLOWED_BLOCKS[ blockName ];
+}
+
+/**
+ * Add custom controls to all blocks
+ */
+const withCustomControls = createHigherOrderComponent( ( BlockEdit ) => {
+	return ( props ) => {
+		const bindableAttributes = getBindableAttributes( props.name );
+		const { updateBlockBindings, removeAllBlockBindings } =
+			useBlockBindingsUtils();
+
+		const { postType, postId } = useSelect( ( select ) => {
+			const { getCurrentPostType, getCurrentPostId } =
+				select( editorStore );
+			return {
+				postType: getCurrentPostType(),
+				postId: getCurrentPostId(),
+			};
+		}, [] );
+
+		const fieldsGroups = useSelect(
+			( select ) => {
+				const { getEditedEntityRecord } = select( coreDataStore );
+
+				if ( ! postType || ! postId ) {
+					return undefined;
+				}
+
+				const record = getEditedEntityRecord(
+					'postType',
+					postType,
+					postId
+				);
+				return record?.scf_field_groups;
+			},
+			[ postType, postId ]
+		);
+
+		const currentBindings = props.attributes?.metadata?.bindings || {};
+
+		const fields = useMemo(
+			() =>
+				fieldsGroups?.reduce( ( acc, fieldGroup ) => {
+					const groupFields =
+						fieldGroup.fields?.map( ( field ) => ( {
+							...field,
+							fieldGroupTitle: fieldGroup.title,
+							name: field.name,
+							label: field.label,
+							value: field.value,
+						} ) ) || [];
+
+					return [ ...acc, ...groupFields ];
+				}, [] ) || [],
+			[ fieldsGroups ]
+		);
```

Comment:
If we don't need `fieldGroupTitle` (see https://github.com/WordPress/secure-custom-fields/pull/176/files#r2197468409), then I think we can simplify this as

```suggestion
		const fields = useMemo(
			() =>
				fieldsGroups.flatMap( ( fieldGroup ) => fieldGroup.fields ) ||
				[],
			[ fieldsGroups ]
		);
```

Reply:
Sorry, the code got messed up, refactored in https://github.com/WordPress/secure-custom-fields/commit/866bb43d399579ab4b23fa1549b3191714f606bc